dattri/algorithm/base.py

```diff
@@ -117,104 +109,135 @@ def _set_full_train_data(self, dataloader: torch.utils.data.DataLoader) -> None:
+                Typically, it is a 2-d dimensional tensor with the shape of
+                (batch_size, num_parameters).
+
+        Returns:
+            torch.Tensor: The transformed test representations. Typically,
+                it is a 2-d dimensional tensor with the shape of
+                (batch_size, transformed_dimension).
+        """
+        return test_rep
 
-        Inner product attributor calculates the inner product between the
-        train query and the transformation of the test query. This function
-        calculates the transformation of the test query.
+    def transform_train_rep(  # noqa: PLR6301
+        self,
+        ckpt_idx: int,  # noqa:ARG002
+        train_rep: torch.Tensor,
+    ) -> torch.Tensor:
+        """Transform the train representations.
 
-        Normally speaking, this function may return any transformation on the query.
-        Hessian Matrix and Fisher Information Matrix are two common transformations.
+        Inner product attributor calculates the inner product between the (transformed)
+        train representations and test representations. This function calculates the
+        transformation of the train representations. For example, the transformation
+        could be a dimension reduction of the train representations.
 
         Args:
-            index (int): The index of the model parameters. This index
-                is used for ensembling of different trained model.
-            train_data (Tuple[Tensor]): Normally this is a tuple
-                of input data and target data, the number of items in the
-                tuple should be aligned in the target function. The tensors'
-                shape follows (1, batchsize, ...).
-            query (torch.Tensor): The query to be transformed. Normally it is
-                a 2-d dimensional tensor with the shape of
-                (batchsize, num_parameters).
+            ckpt_idx (int): The index of the model checkpoints. This index
+                is used for ensembling different trained model checkpoints.
+            train_rep (torch.Tensor): The train representations to be transformed.
+                Typically, it is a 2-d dimensional tensor with the shape of
+                (batch_size, num_parameters).
 
         Returns:
-            torch.Tensor: The transformation on the query. Normally it is
-                a 2-d dimensional tensor with the shape of
-                (batchsize, transformed_dimension).
+            torch.Tensor: The transformed train representations. Typically,
+                it is a 2-d dimensional tensor with the shape of
+                (batch_size, transformed_dimension).
         """
+        return train_rep
 
-    def cache(self, full_train_dataloader: DataLoader) -> None:
-        """Cache the dataset for inverse hessian calculation.
+    def cache(self, full_train_dataloader: torch.utils.data.DataLoader) -> None:
+        """Cache the full training dataloader.
```

Comment:
or mention some pre-computation  can also be cached?
since it can be more things